test/parse_helper.rb

```diff
@@ -273,7 +282,7 @@ def parse_source_map_descriptions(descriptions)
 
       if (match = SOURCE_MAP_DESCRIPTION_RE.match(line))
         begin_pos        = match[1].length
-        end_pos          = begin_pos + match[2].length
+        end_pos          = begin_pos + match[2].length unless match[2] == '!'
```

Comment:
1. the pattern allows `!!!! begin (ast_path)` that makes no sense. `!` does not represent any source range
2. this very special `if` feels like a hack
3. the caller checks `end_pos`, but there's also `begin_pos` and it's unclear why only one of them is important

What do you think about yielding something like `source_map_type` (as the first argument) that is either `:assert_source_map` or `:assert_no_source_map` (not the best naming, but I hope you get the idea). This way you could do `case source_map_type ...` on the call site to perform either `assert_source_range(...)` or `assert_no_source_range(...)`

Reply:
Most definitely a hack 😅